CorsixTH/Src/iso_fs.cpp

```diff
@@ -260,17 +275,32 @@ class iso_directory_iterator final {
     /**
      * Advance this iterator to the next file entry in the directory table,
      * returning the result.
+     * In cases where advancing the iterator would read past the end of the
+     * directory table, an exception is thrown and the iterator is not
+     * advanced.
      */
     iso_directory_iterator& operator++() {
-        directory_ptr += *directory_ptr;
-        while (*directory_ptr == 0 && directory_ptr < end_ptr) {
-            ++directory_ptr;
+        if (directory_ptr >= end_ptr) {
+            throw std::out_of_range("Cannot advance iso directory iterator past end of input");
         }
-        if (directory_ptr == end_ptr) {
-            entry = iso_file_entry();
+
+        const uint8_t* new_dir_ptr = directory_ptr + *directory_ptr;
+        while (new_dir_ptr < end_ptr && *new_dir_ptr == 0) {
+            ++new_dir_ptr;
+        }
+
+        // Catch a malformed directory entry where the size would extend past
+        // the end of the table.
+        if (new_dir_ptr < end_ptr && new_dir_ptr + *new_dir_ptr > end_ptr) {
+            throw std::runtime_error("The last directory entry was larger than the defined table region.");
+        }
+
+        if (new_dir_ptr < end_ptr) {
```

Comment:
Since the first check at line 294 is the same, you can move the comments, the second check, and the throw into here.

Reply:
I felt it was clearer this way.